libcache/cache.c

```diff
@@ -0,0 +1,460 @@
+		free(cache);
+		return NULL;
+	}
+
+	cache->lineSize = lineSize;
+	cache->numLines = size / cache->lineSize;
+
+	cache->numSets = cache->numLines / LIBCACHE_NUM_WAYS;
+
+	cache->sets = calloc(cache->numSets, sizeof(cacheset_t));
+
+	if (cache->sets == NULL) {
+		free(cache);
+		return NULL;
+	}
+
+	for (; i < cache->numSets; ++i) {
+		cache_setInit(&cache->sets[i]);
+	}
+
+	cache->offBitNum = log2(cache->lineSize);
+	cache->setBitsNum = log2(cache->numSets);
+	cache->tagBitsNum = LIBCACHE_ADDR_WIDTH - cache->setBitsNum - cache->offBitNum;
+
+	cache->tagMask = cache_genMask(cache->tagBitsNum);
+	cache->setMask = cache_genMask(cache->setBitsNum);
+	cache->offMask = cache_genMask(cache->offBitNum);
+
+	cache->readCb = readCb;
+	cache->writeCb = writeCb;
+
+	return cache;
+}
+
+
+int cache_deinit(cachectx_t *cache)
+{
+	int i, j;
+
+	for (i = 0; i < cache->numSets; ++i) {
+		for (j = 0; j < LIBCACHE_NUM_WAYS; ++j) {
+			if (IS_VALID(cache->sets[i].lines[j].flags)) {
+				free(cache->sets[i].lines[j].data);
+				CLEAR_VALID(cache->sets[i].lines[j].flags);
+			}
+		}
+	}
+	free(cache->sets);
+
+	free(cache);
+
+	return 0;
+}
+
+
+static int cache_compareTags(const void *lhs, const void *rhs)
+{
+	int ret = 0;
+
+	cacheline_t **lhsLine = (cacheline_t **)lhs;
+	cacheline_t **rhsLine = (cacheline_t **)rhs;
+
+	if (*lhsLine == NULL && *rhsLine == NULL) {
+		ret = 0;
+	}
+	else if (*lhsLine == NULL && *rhsLine != NULL) {
+		ret = -1;
+	}
+	else if (*lhsLine != NULL && *rhsLine == NULL) {
+		ret = 1;
+	}
+	else {
+		if ((*lhsLine)->tag < (*rhsLine)->tag) {
+			ret = -1;
+		}
+		else if ((*lhsLine)->tag == (*rhsLine)->tag) {
+			ret = 0;
+		}
+		else {
+			ret = 1;
+		}
+	}
+
+	return ret;
+}
+
+
+static ssize_t cache_writeToSet(cache_writeCb_t writeCb, cacheset_t *set, cacheline_t *line, const uint64_t offset, size_t count, int policy)
+{
+	int i, j;
+	ssize_t written = 0;
+	cacheline_t *temp = NULL, **linePtr = NULL;
+
+	if (set->count < LIBCACHE_NUM_WAYS) {
+		for (i = 0; i < LIBCACHE_NUM_WAYS; ++i) {
+			if (!IS_VALID(set->lines[i].flags)) {
+				set->lines[i] = *line;
+				break;
+			}
+		}
+
+		temp = &(set->lines[i]);
+
+		for (j = 0; j < LIBCACHE_NUM_WAYS; ++j) {
+			if (set->tags[j] == NULL) {
+				set->tags[j] = temp;
+				break;
+			}
+		}
+
+		LIST_ADD(&set->timestamps, temp);
+		set->count++;
+	}
+	else {
+		temp = set->timestamps;
+		LIST_REMOVE(&set->timestamps, temp);
+
+		/* write-back */
+		if (IS_DIRTY(temp->flags)) {
+			written = (*writeCb)(offset, temp->data, sizeof(*temp->data), count);
+			free(temp->data);
+		}
+
+		for (i = 0; i < LIBCACHE_NUM_WAYS; ++i) {
+			if (IS_VALID(set->lines[i].flags) && (set->lines[i].tag == temp->tag)) {
+				set->lines[i] = *line;
+				break;
+			}
+		}
+
+		LIST_ADD(&set->timestamps, temp);
+	}
+
+	switch (policy) {
+		case 0:
+			/* write-back */
+			SET_DIRTY(set->lines[i].flags);
```

Comment:
What if `i == LIBCACHE_NUM_WAYS`

Reply:
I think ``i`` should not be incremented beyond ``LIBCACHE_NUM_WAYS``. Have I missed something?